src/BindPattern.jl

```diff
@@ -500,33 +500,43 @@ function shred_where_clause(
         result_type = (inverted == (guard.head == :&&)) ? BoundOrPattern : BoundAndPattern
         return result_type(location, guard, BoundPattern[left, right])
     else
-        (guard0, assigned0) = subst_patvars(guard, assigned)
-        fetch = BoundFetchExpressionPattern(location, guard, guard0, assigned0, nothing, Any)
+        bound_expression = bind_expression(location, guard, assigned)
+        fetch = BoundFetchExpressionPattern(bound_expression, nothing, Any)
         temp = get_temp(binder, fetch)
         test = BoundWhereTestPattern(location, guard, temp, inverted)
         return BoundAndPattern(location, guard, BoundPattern[fetch, test])
     end
 end
 
 #
-# Replace each pattern variable reference with the temporary variable holding the
-# value that corresponds to that pattern variable.
+# Produce a `BoundExpression` object for the given expression.  This is used to
+# determine the set of variable bindings that are used in the expression, and to
+# simplify code generation.
 #
-function subst_patvars(expr, assigned::ImmutableDict{Symbol, Symbol})
-    new_assigned = ImmutableDict{Symbol, Symbol}()
-    new_expr = MacroTools.postwalk(expr) do patvar
-        if patvar isa Symbol
-            tmpvar = get(assigned, patvar, nothing)
-            if tmpvar isa Symbol
-                if !haskey(new_assigned, patvar)
-                    new_assigned = ImmutableDict{Symbol, Symbol}(
-                        new_assigned, patvar, tmpvar)
-                end
-                # Prevent the variable from being assigned to in user code
-                return Expr(:block, tmpvar)
-            end
-        end
+function bind_expression(location::LineNumberNode, expr, assigned::ImmutableDict{Symbol, Symbol})
+    if is_expr(expr, :(...))
+        error("$(location.file):$(location.line): Splatting not supported in interpolation: `$expr`.")
+    end
+
+    # determine the variables *actually* used in the expression
+    used = Set{Symbol}()
+    MacroTools.postwalk(expr) do patvar
+        patvar isa Symbol && push!(used, patvar)
         patvar
     end
-    (new_expr, new_assigned)
+
+    # we sort the used variables by name so that we have a deterministic order
+    # that will make it more likely we can share the resulting expression.
+    used = sort(collect(intersect(keys(assigned), used)))
+
+    assignments = Expr(:block)
+    new_assigned = ImmutableDict{Symbol, Symbol}()
+    for v in used
+        tmp = get(assigned, v, nothing)
+        @assert tmp !== nothing
```

Comment:
what if a plain/explicit symbol were used in the expression, would this assertion still hold?

Reply:
Yes.  `used` is the intersection of variable names appearing in the expression and declared pattern variable names (line 530), so when it finds a variable that was in used, it must have been a pattern variable name and therefore in the map.